File: packages/app-game-domain/src/app-game-notification-parent-surface-intent.ts
import {
  AppGameNotificationPreferenceStatusHandoffReadModelSchema,
} from '@ocentra-parent/schema-domain/app-game-notification-preference-status-handoff';
import type {
  AppGameNotificationPreferenceStatusHandoffReadModel,
  AppGameNotificationPreferenceStatusHandoffRow,
} from '@ocentra-parent/schema-domain/app-game-notification-preference-status-handoff';
import {
  AppGameNotificationProviderStatusHandoffReadModelSchema,
} from '@ocentra-parent/schema-domain/app-game-notification-provider-status-handoff';
import type {
  AppGameNotificationProviderStatusHandoffReadModel,
  AppGameNotificationProviderStatusHandoffRow,
} from '@ocentra-parent/schema-domain/app-game-notification-provider-status-handoff';
import {
  AppGameNotificationParentSurfaceIntentReadModelSchema,
  AppGameNotificationParentSurfaceIntentRowSchema,
  RequiredAppGameNotificationParentSurfaceIntentNonClaims,
  appGameNotificationParentSurfaceHistoryVisibilityFor,
} from '@ocentra-parent/schema-domain/app-game-notification-parent-surface-intent';
import type {
  AppGameNotificationParentSurfaceIntentReadModel,
  AppGameNotificationParentSurfaceIntentRow,
} from '@ocentra-parent/schema-domain/app-game-notification-parent-surface-intent';
import { ParentContractSchemaVersion } from '@ocentra-parent/schema-domain/family-reference-primitives';

export type AppGameNotificationParentSurfaceIntentOptions = {
  readonly generatedAt: string;
  readonly intentId: string;
  readonly sourceContractRefs: readonly string[];
};

export function buildAppGameNotificationParentSurfaceIntentReadModel(
  options: AppGameNotificationParentSurfaceIntentOptions,
  providerReadModel: AppGameNotificationProviderStatusHandoffReadModel,
  preferenceReadModel: AppGameNotificationPreferenceStatusHandoffReadModel
): AppGameNotificationParentSurfaceIntentReadModel {
  const parsedProvider = AppGameNotificationProviderStatusHandoffReadModelSchema.parse(providerReadModel);
  const parsedPreference = AppGameNotificationPreferenceStatusHandoffReadModelSchema.parse(preferenceReadModel);
  assertCompatibleInputs(parsedProvider, parsedPreference);

  const rows = parsedProvider.rows.map((providerRow, index) =>
    parentSurfaceIntentRowForStatusRows(providerRow, preferenceRowAt(parsedPreference, index))
  );

  return AppGameNotificationParentSurfaceIntentReadModelSchema.parse({
    schemaVersion: ParentContractSchemaVersion.V0_6,
    intentId: options.intentId,
    generatedAt: options.generatedAt,
    family: parsedProvider.family,
    sourceProviderStatusHandoffId: parsedProvider.handoffId,
    sourcePreferenceStatusHandoffId: parsedPreference.handoffId,
    sourceContractRefs: options.sourceContractRefs,
    rows,
    manualActionRequiredCount: rows.filter((row) => row.parentSurfaceStatus === 'manual-action-required').length,
    unavailableVisibleCount: rows.filter((row) => row.parentSurfaceStatus === 'unavailable-visible').length,
    historyVisibleCount: rows.length,
    preferenceSetupRequiredCount: rows.filter((row) => row.preferenceVisibility === 'preference-setup-required').length,
    parentSurfaceNonClaims: RequiredAppGameNotificationParentSurfaceIntentNonClaims,
    parentNotificationUiRendered: false,
    parentPreferenceUiRendered: false,
    parentFrequencyControlUiRendered: false,
    providerDeliveryRuntimeClaimed: false,
    providerReceiptIngestionClaimed: false,
    providerCredentialsClaimed: false,
    cloudRoutingClaimed: false,
    childDeliveryClaimed: false,
    productionRuntimeClaimed: false,
    productionDurableOutboxStorageClaimed: false,
    adapterDispatchClaimed: false,
  });
}

function assertCompatibleInputs(
  providerReadModel: AppGameNotificationProviderStatusHandoffReadModel,
  preferenceReadModel: AppGameNotificationPreferenceStatusHandoffReadModel
): void {
  if (providerReadModel.family.familyId !== preferenceReadModel.family.familyId) {
    throw new Error('Expected app/game notification parent-surface inputs to use the same family ref');
  }
  if (providerReadModel.rows.length !== preferenceReadModel.rows.length) {
    throw new Error('Expected app/game notification parent-surface inputs to have matching row counts');
  }
}

function preferenceRowAt(
  preferenceReadModel: AppGameNotificationPreferenceStatusHandoffReadModel,
  index: number
): AppGameNotificationPreferenceStatusHandoffRow {
  const preferenceRow = preferenceReadModel.rows[index];
  if (preferenceRow === undefined) {
    throw new Error('Expected app/game notification parent-surface preference row to exist');
  }
  return preferenceRow;
}

function parentSurfaceIntentRowForStatusRows(
  providerRow: AppGameNotificationProviderStatusHandoffRow,
  preferenceRow: AppGameNotificationPreferenceStatusHandoffRow
): AppGameNotificationParentSurfaceIntentRow {
  const providerEntry = providerRow.providerStatusBoundaryEntry;
  const preferenceEntry = preferenceRow.notificationPreferenceStatusEntry;

  return AppGameNotificationParentSurfaceIntentRowSchema.parse({
    surfaceRowId: `app-game-notification-parent-surface-${providerRow.handoffRowId}`,
    sourceProviderHandoffRowId: providerRow.handoffRowId,
    sourcePreferenceHandoffRowId: preferenceRow.handoffRowId,
    sourceSchedulerEntryRef: providerRow.sourceSchedulerEntryRef ?? preferenceRow.sourceSchedulerEntryRef,
    sourceOutboxRecordRef: providerRow.sourceOutboxRecordRef ?? preferenceRow.sourceOutboxRecordRef,
    providerStatus: providerEntry.providerStatus,
    deliveryResultState: preferenceEntry.deliveryResultState,
    parentPreferenceState: preferenceEntry.parentPreferenceState,
    quietHoursDecision: preferenceEntry.quietHoursDecision,
    providerChannel: preferenceEntry.providerChannel,
    parentSurfaceStatus:
      providerEntry.providerStatus === 'unavailable' ? 'unavailable-visible' : 'manual-action-required',
    historyVisibility: appGameNotificationParentSurfaceHistoryVisibilityFor(providerEntry.providerStatus),
    preferenceVisibility:
      preferenceEntry.parentPreferenceState === 'channel-disabled'
        ? 'preference-disabled-visible'
        : 'preference-setup-required',
    drillInRefs: [providerEntry.notificationStatusRef, preferenceEntry.deliveryResultRef],
    auditRefs: [...providerEntry.auditRefs, ...preferenceEntry.auditRefs],
    manualProofRequirements: [...providerEntry.manualProofRequirements, ...preferenceEntry.manualProofRequirements],
    minimalSurfacePayloadBoundary:
      'Parent surface intent contains status refs and setup requirements only; sensitive app/game evidence stays behind authenticated drill-in.',
    sensitiveDetailIncluded: false,
    providerDeliveryClaimed: false,
    providerReceiptClaimed: false,
    parentPreferenceMutationClaimed: false,
    childDeliveryClaimed: false,
  });
}

File: crates/agent-protocol/src/app_game_notification_parent_surface_intent.rs
use serde::{Deserialize, Serialize};

pub const APP_GAME_NOTIFICATION_PARENT_SURFACE_INPUT_FAMILY_MISMATCH: &str =
    "Expected app/game notification parent-surface inputs to use the same family ref";
pub const APP_GAME_NOTIFICATION_PARENT_SURFACE_INPUT_ROW_COUNT_MISMATCH: &str =
    "Expected app/game notification parent-surface inputs to have matching row counts";

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGameNotificationParentSurfaceIntentOptions {
    pub generated_at: String,
    pub intent_id: String,
    pub source_contract_refs: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGameNotificationFamilyReference {
    pub family_id: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGameNotificationProviderStatusBoundaryEntry {
    pub provider_status: String,
    pub notification_status_ref: String,
    pub audit_refs: Vec<String>,
    pub manual_proof_requirements: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGameNotificationProviderStatusHandoffRow {
    pub handoff_row_id: String,
    pub source_scheduler_entry_ref: Option<String>,
    pub source_outbox_record_ref: Option<String>,
    pub provider_status_boundary_entry: AppGameNotificationProviderStatusBoundaryEntry,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGameNotificationProviderStatusHandoffReadModel {
    pub handoff_id: String,
    pub family: AppGameNotificationFamilyReference,
    pub rows: Vec<AppGameNotificationProviderStatusHandoffRow>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGameNotificationPreferenceStatusHandoffRow {
    pub handoff_row_id: String,
    pub source_scheduler_entry_ref: Option<String>,
    pub source_outbox_record_ref: Option<String>,
    pub notification_preference_status_entry: AppGameNotificationPreferenceStatusHandoffEntry,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGameNotificationPreferenceStatusHandoffEntry {
    pub delivery_result_state: String,
    pub parent_preference_state: String,
    pub quiet_hours_decision: String,
    pub provider_channel: String,
    pub delivery_result_ref: String,
    pub audit_refs: Vec<String>,
    pub manual_proof_requirements: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGameNotificationPreferenceStatusHandoffReadModel {
    pub handoff_id: String,
    pub family: AppGameNotificationFamilyReference,
    pub rows: Vec<AppGameNotificationPreferenceStatusHandoffRow>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGameNotificationParentSurfaceIntentRow {
    pub surface_row_id: String,
    pub source_provider_handoff_row_id: String,
    pub source_preference_handoff_row_id: String,
    pub source_scheduler_entry_ref: Option<String>,
    pub source_outbox_record_ref: Option<String>,
    pub provider_status: String,
    pub delivery_result_state: String,
    pub parent_preference_state: String,
    pub quiet_hours_decision: String,
    pub provider_channel: String,
    pub parent_surface_status: String,
    pub history_visibility: String,
    pub preference_visibility: String,
    pub drill_in_refs: Vec<String>,
    pub audit_refs: Vec<String>,
    pub manual_proof_requirements: Vec<String>,
    pub minimal_surface_payload_boundary: String,
    pub sensitive_detail_included: bool,
    pub provider_delivery_claimed: bool,
    pub provider_receipt_claimed: bool,
    pub parent_preference_mutation_claimed: bool,
    pub child_delivery_claimed: bool,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGameNotificationParentSurfaceIntentReadModel {
    pub schema_version: String,
    pub intent_id: String,
    pub generated_at: String,
    pub family: AppGameNotificationFamilyReference,
    pub source_provider_status_handoff_id: String,
    pub source_preference_status_handoff_id: String,
    pub source_contract_refs: Vec<String>,
    pub rows: Vec<AppGameNotificationParentSurfaceIntentRow>,
    pub manual_action_required_count: usize,
    pub unavailable_visible_count: usize,
    pub history_visible_count: usize,
    pub preference_setup_required_count: usize,
    pub parent_surface_non_claims: Vec<String>,
    pub parent_notification_ui_rendered: bool,
    pub parent_preference_ui_rendered: bool,
    pub parent_frequency_control_ui_rendered: bool,
    pub provider_delivery_runtime_claimed: bool,
    pub provider_receipt_ingestion_claimed: bool,
    pub provider_credentials_claimed: bool,
    pub cloud_routing_claimed: bool,
    pub child_delivery_claimed: bool,
    pub production_runtime_claimed: bool,
    pub production_durable_outbox_storage_claimed: bool,
    pub adapter_dispatch_claimed: bool,
}
